File: backend/app/schemas/menu.py
import uuid

from pydantic import BaseModel, Field


class MenuCategoryCreate(BaseModel):
    name: str = Field(min_length=1, max_length=150)
    description: str | None = None
    display_order: int = 0


class MenuCategoryUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=150)
    description: str | None = None
    display_order: int | None = None
    is_active: bool | None = None


class MenuCategoryOut(BaseModel):
    id: uuid.UUID
    name: str
    description: str | None
    display_order: int
    is_active: bool

    model_config = {"from_attributes": True}


class MenuOptionCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    price_delta: float = 0
    display_order: int = 0


class MenuOptionUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    price_delta: float | None = None
    display_order: int | None = None
    is_active: bool | None = None


class MenuOptionOut(BaseModel):
    id: uuid.UUID
    name: str
    price_delta: float
    is_active: bool
    display_order: int

    model_config = {"from_attributes": True}


class MenuOptionGroupCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    is_required: bool = False
    allow_multiple: bool = False
    display_order: int = 0
    options: list[MenuOptionCreate] = Field(default_factory=list)


class MenuOptionGroupUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    is_required: bool | None = None
    allow_multiple: bool | None = None
    display_order: int | None = None
    is_active: bool | None = None


class MenuOptionGroupOut(BaseModel):
    id: uuid.UUID
    name: str
    is_required: bool
    allow_multiple: bool
    display_order: int
    is_active: bool
    options: list[MenuOptionOut] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class MenuVariantCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    price_delta: float = 0
    is_default: bool = False


class MenuVariantUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    price_delta: float | None = None
    is_default: bool | None = None
    is_active: bool | None = None


class MenuVariantOut(BaseModel):
    id: uuid.UUID
    name: str
    price_delta: float
    is_default: bool
    is_active: bool

    model_config = {"from_attributes": True}


class MenuItemCreate(BaseModel):
    category_id: uuid.UUID
    kitchen_station: str | None = Field(default=None, max_length=50)
    name: str = Field(min_length=1, max_length=200)
    description: str | None = None
    base_price: float = Field(gt=0)
    is_veg: bool = True
    image_url: str | None = None
    display_order: int = 0
    variants: list[MenuVariantCreate] = Field(default_factory=list)
    option_groups: list[MenuOptionGroupCreate] = Field(default_factory=list)


class MenuItemUpdate(BaseModel):
    category_id: uuid.UUID | None = None
    kitchen_station: str | None = Field(default=None, max_length=50)
    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None
    base_price: float | None = Field(default=None, gt=0)
    is_veg: bool | None = None
    image_url: str | None = None
    display_order: int | None = None
    is_active: bool | None = None
    is_sold_out: bool | None = None
    is_todays_special: bool | None = None
    is_specialty: bool | None = None


class MenuItemOut(BaseModel):
    id: uuid.UUID
    category_id: uuid.UUID
    kitchen_station: str | None = None
    name: str
    description: str | None
    base_price: float
    is_veg: bool
    is_active: bool
    is_sold_out: bool
    is_todays_special: bool
    is_specialty: bool
    image_url: str | None
    display_order: int
    variants: list[MenuVariantOut] = Field(default_factory=list)
    option_groups: list[MenuOptionGroupOut] = Field(default_factory=list)

    model_config = {"from_attributes": True}

File: frontend/src/api/menu.ts
import { apiClient } from "@/api/client";
import { normalizeMenuItem } from "@/api/normalize";
import type {
  MenuCategoryCreate,
  MenuCategoryOut,
  MenuCategoryUpdate,
  MenuItemCreate,
  MenuItemOut,
  MenuItemUpdate,
  MenuOptionCreate,
  MenuOptionGroupCreate,
  MenuOptionGroupOut,
  MenuOptionGroupUpdate,
  MenuOptionOut,
  MenuOptionUpdate,
  MenuVariantCreate,
  MenuVariantOut,
  MenuVariantUpdate,
} from "@/types/models";

export const categoriesApi = {
  list: (signal?: AbortSignal) =>
    apiClient.get<MenuCategoryOut[]>("/categories", { signal }).then((r) => r.data),

  create: (payload: MenuCategoryCreate) =>
    apiClient.post<MenuCategoryOut>("/categories", payload).then((r) => r.data),

  update: (categoryId: string, payload: MenuCategoryUpdate) =>
    apiClient.put<MenuCategoryOut>(`/categories/${categoryId}`, payload).then((r) => r.data),

  delete: (categoryId: string) => apiClient.delete(`/categories/${categoryId}`),
};

export const menuApi = {
  listItems: (params: { category_id?: string } = {}, signal?: AbortSignal) =>
    apiClient
      .get<MenuItemOut[]>("/menu/items", { params, signal })
      .then((r) => r.data.map(normalizeMenuItem)),

  createItem: (payload: MenuItemCreate) =>
    apiClient.post<MenuItemOut>("/menu/items", payload).then((r) => normalizeMenuItem(r.data)),

  updateItem: (itemId: string, payload: MenuItemUpdate) =>
    apiClient.put<MenuItemOut>(`/menu/items/${itemId}`, payload).then((r) => normalizeMenuItem(r.data)),

  deleteItem: (itemId: string) => apiClient.delete(`/menu/items/${itemId}`),

  addVariant: (itemId: string, payload: MenuVariantCreate) =>
    apiClient.post<MenuItemOut>(`/menu/items/${itemId}/variants`, payload).then((r) => normalizeMenuItem(r.data)),

  addOptionGroup: (itemId: string, payload: MenuOptionGroupCreate) =>
    apiClient.post<MenuItemOut>(`/menu/items/${itemId}/option-groups`, payload).then((r) => normalizeMenuItem(r.data)),

  addOption: (groupId: string, payload: MenuOptionCreate) =>
    apiClient.post<MenuOptionOut>(`/menu/option-groups/${groupId}/options`, payload).then((r) => r.data),

  updateVariant: (variantId: string, payload: MenuVariantUpdate) =>
    apiClient.put<MenuVariantOut>(`/menu/variants/${variantId}`, payload).then((r) => r.data),

  deleteVariant: (variantId: string) => apiClient.delete(`/menu/variants/${variantId}`),

  updateOptionGroup: (groupId: string, payload: MenuOptionGroupUpdate) =>
    apiClient.put<MenuOptionGroupOut>(`/menu/option-groups/${groupId}`, payload).then((r) => r.data),

  deleteOptionGroup: (groupId: string) => apiClient.delete(`/menu/option-groups/${groupId}`),

  updateOption: (optionId: string, payload: MenuOptionUpdate) =>
    apiClient.put<MenuOptionOut>(`/menu/options/${optionId}`, payload).then((r) => r.data),

  deleteOption: (optionId: string) => apiClient.delete(`/menu/options/${optionId}`),
};
